internal/httputil/certpoolwatcher.go

```diff
@@ -0,0 +1,108 @@
+package httputil
+
+import (
+	"crypto/x509"
+	"fmt"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/go-logr/logr"
+)
+
+type CertPoolWatcher struct {
+	generation int
+	dir        string
+	mx         sync.RWMutex
+	pool       *x509.CertPool
+	log        logr.Logger
+	watcher    *fsnotify.Watcher
+	done       chan bool
+}
+
+// Returns the current CertPool and the generation number
+func (cpw *CertPoolWatcher) Get() (*x509.CertPool, int, error) {
+	cpw.mx.RLock()
+	defer cpw.mx.RUnlock()
+	if cpw.pool == nil {
+		return nil, 0, fmt.Errorf("no certificate pool available")
+	}
+	return cpw.pool.Clone(), cpw.generation, nil
+}
+
+func (cpw *CertPoolWatcher) Done() {
+	cpw.done <- true
+}
+
+func NewCertPoolWatcher(caDir string, log logr.Logger) (*CertPoolWatcher, error) {
+	pool, err := NewCertPool(caDir, log)
+	if err != nil {
+		return nil, err
+	}
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+	if err = watcher.Add(caDir); err != nil {
+		return nil, err
+	}
+
+	cpw := &CertPoolWatcher{
+		generation: 1,
+		dir:        caDir,
+		pool:       pool,
+		log:        log,
+		watcher:    watcher,
+		done:       make(chan bool),
+	}
+	go func() {
+		for {
+			select {
+			case <-watcher.Events:
+				cpw.drainEvents()
+				cpw.update()
+			case err := <-watcher.Errors:
+				log.Error(err, "error watching certificate dir")
+				os.Exit(1)
+			case <-cpw.done:
+				err := watcher.Close()
+				if err != nil {
+					log.Error(err, "error closing watcher")
+				}
+				return
+			}
+		}
+	}()
+	return cpw, nil
+}
+
+func (cpw *CertPoolWatcher) update() {
+	cpw.log.Info("updating certificate pool")
+	pool, err := NewCertPool(cpw.dir, cpw.log)
+	if err != nil {
+		cpw.log.Error(err, "error updating certificate pool")
+		os.Exit(1)
```

Comment:
Is it expected that encountering an error when performing an update to the certificate pool that restarting the pod will result in the error being resolved?

Reply:
If we can't create the cert pool, then something is wrong with the pod. The mount of the certs directory is broken, or the set of certificates is bad, or the configuration of the pod is bad. Joe has given a previous comment he'd rather have the pod crash (and be noticed) than have it continue to run in an error condition.